### PR TITLE
@types/vuelidate

### PR DESCRIPTION
**Pull request by janesser**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by janesser** [comment]
Originated from https://github.com/monterail/vuelidate/pull/287

**Comment by janesser** [comment]
Filed to get dependency vue-property-decorator in type-publishers whitelist: https://github.com/Microsoft/types-publisher/issues/504

**Comment by janesser** [comment]
TODO replace @types/vue with vue itself.

File: types/vuelidate/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Type definitions for vuelidate 0.7
+// Project: https://github.com/janesser/vuelidate.ts
+// Definitions by: Jan Esser <https://github.com/janesser>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.4
+
+import "./vue"
```
Comment (Rouche) [inline types/vuelidate/index.d.ts]: single quotes

File: types/vuelidate/vue.d.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// vue augmentation
+import Vue, { ComponentOptions } from 'vue'
+
+import { ValidationRule } from './lib/validators'
+import { Validation } from "./vuelidate"
```
Comment (Rouche) [inline types/vuelidate/vue.d.ts]: same here.

File: types/vuelidate/lib/validators.d.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+// const ValidationRule
+export interface ValidationRule {
+    $params(): ValidationParams
+    $pending(): boolean
+}
+
+export type CustomRule = (value: any, parentVm?: Vue) => boolean
+
+export interface Helpers {
```
Comment (Rouche) [inline types/vuelidate/lib/validators.d.ts]:
```
    export interface Helpers {
        withParams(params: Params, rule: CustomRule): ValidationRule;
        req(field: string): ValidationRule;
        ref(reference: any, vm: Vue, parentVm: Vue): any;
        len(value: any): number;
        regex(type: any, expr: RegExp): ValidationRule;
    }
```

**Comment by janesser** [comment]
Replaced deprecated "@types/vue" with "vue/types".

Completed helpers, covered by type-tests.

**Comment by janesser** [comment]
The README.md does give hints about how test the typing in the wild.

I tried a softlink from the node_modules of my vue-playground test-project. This won't resolve @types/vuelidate to vue dependency.

Also tried the same with having vue in node_modules of @types/vuelidate which rsults in two concurrent and incompatible versions of vue.

Hints welcome!

**Comment by Rouche** [comment]
@janesser Im also stuck. TypeScript compiler is never using @types/vulidate for main index.d.ts

It always gets the index.js from vuelidate. That is very weird.

I really have no idea why. I dont see differences between vuelidate and @types/tinymce that works perfectly.

Maybe you can check why tinymce type works and vuelidate refuse to cooperate :)

File: types/vuelidate/tslint.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
+        "semicolon": false,
+        "strict-export-declare-modifiers": false,
+        "no-single-declare-module": false
```
Comment (DanielRosenwasser) [inline types/vuelidate/tslint.json]: You shouldn't need this since those are module augmentations (i.e. they're declared inside of a module file).

Comment (DanielRosenwasser) [reply types/vuelidate/tslint.json]: Also, for the sake of repo-wide consistency, turning the other rules off would be appreciated (though isn't strictly required).

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by clarenced** [comment]
@janesser Any update on this ?

**Comment by janesser** [comment]
> 
> 
> @janesser Any update on this PR ?

will come back here on the week-end.

**Comment by janesser** [comment]
@DanielRosenwasser i reduced the tslint.json. Let me know if i should consider further adjustments.

**Pull request merged**